### PR TITLE
[MWES-2871] Replace Drupal database caching with memcached

### PR DESCRIPTION
**Pull request by robpblake**
This PR sets the ground-work for replacing Drupals default database caching with memcached. The following changes are introduced in this PR:

* Updated our image to use PHP 7.2 and installed the [memcached PECL](https://pecl.php.net/package/memcached)
* Installed the latest version of the memcached module in Drupal
* Extended the `redhat_developer` configuration object to include the `cache.engine` key. When this is set to `memcached`, then memcached is used as the cache engine of choice. By default this remains the database.
* Updated the local development experience to provide support for running memcached locally.

This PR does not yet enable memcached as the cache engine of choice. This is because we need to install the module first via `drush cim` and then switch the cache engine to be memcached. In addition I want to perform the migration to memcached in preview environments first to ensure it works as expected.

A follow up PR will enable memcached in preview environments and then prod.

### JIRA Issue Link
* https://issues.jboss.org/browse/MWES-2871

### Verification Process

* The build must go green
* Log in to Drupal and ensure that there is no dangling configuration
* Log in to Drupal and visit `/admin/reports/memcache` and ensure that it reports no memcached instances as in use
* Visit `/admin/reports/status` and ensure that it reports PHP as having `256M` of memory
* Log in to the container and verify that the following hosts are present in the `rhd.settings.php` as `memcached` cache nodes:
  * memcached-0.memcached-headless.rhdp--preview.svc.cluster.local
  * memcached-1.memcached-headless.rhdp--preview.svc.cluster.local
* [Log in to the container](https://github.com/redhat-developer/developers.redhat.com/tree/master/previews) and ensure that php version is `7.2.10`
  * run `php --version` from the command line
* [Log in to the container](https://github.com/redhat-developer/developers.redhat.com/tree/master/previews) and:
  * Edit `rhd.settings.php` and change the value of `$config['redhat_developer']['cache']['engine']` to `memcached`
  * `drush cr` from the command line
  * Revisit `/admin/reports/memcache` and notice now that there are two memcached nodes in-use
  * Browse the site and ensure things work as expected
* Visit the [Grafana Dashboard](https://grafana.util.phx2.redhat.com/d/XYrGhvDZk/developer-drupal?orgId=1), select Drupal Environment `preview` and ensure that metrics are being reported for both `memcached` instances

### CODE REVIEW

File: _docker/drupal/managed-platform/ansible/templates/dev/drupal/rhd.settings.php.j2 [hunk 1]
```diff
@@ -106,4 +105,104 @@ $config['rhd_disqus.disqussettings']['rhd_disqus_secret_key'] = "{{ drupal_disqu
...
+    */
+    $settings['container_yamls'][] = DRUPAL_ROOT . '/sites/default/memcached.services.yml';
+
+    /*
+        Taken from: https://git.drupalcode.org/project/memcache/blob/8.x-2.x/README.txt
+    */
+
+    $class_loader->addPsr4('Drupal\\memcache\\', 'modules/contrib/memcache/src');
+    // Define custom bootstrap container definition to use Memcache for cache.container.
+    $settings['bootstrap_container_definition'] = [
```
Comment (robpblake) [inline _docker/drupal/managed-platform/ansible/templates/dev/drupal/rhd.settings.php.j2]: @fearlsgroove @jordanpagewhite Could you have a look at this part for me. I've copied this verbatim from the memcached module README.txt. I'm not totally up on the implications of this so just would like someone with better Drupal experience than I to +1 it.

cheers,

Rob

Comment (jordanpagewhite) [reply _docker/drupal/managed-platform/ansible/templates/dev/drupal/rhd.settings.php.j2]: Hey @robpblake . This looks good to me after reviewing what you have here and the module. I am not very familiar with configuring Memcache, particularly in recent history as I almost exclusively worked with Paas platforms at my last workplace, but I think it is fair to trust the docs of the Memcache module as the usage is so high, it has an active issue queue, and it is supported by 2 well-known Drupal organizations.

Comment (robpblake) [reply _docker/drupal/managed-platform/ansible/templates/dev/drupal/rhd.settings.php.j2]: @jordanpagewhite I'm more interested in what is happening to the Drupal container definition in `rhd.settings.php` when we have `memcached` enabled. If you're comfortable with it, then so am I. I just don't want to blindly apply things without really having a true grasp of what it is doing.

Lemme know if you're happy.

cheers,

Rob

**Comment by robpblake** [comment]
rebuild this please

**Comment by robpblake** [comment]
rebuild this please

**Comment by robpblake** [comment]
@jordanpagewhite / @fearlsgroove This is ready for review. Please do not merge until it without letting me know as I need to co-ordinate some changes into our production environment beforehand.

cheers,

Rob

**Comment by jordanpagewhite** [comment]
I have verified the following:

* No dangling config
* No unexpected errors reported on the Status page
* Build is green
* No memcached instances reported in use on admin Memcache page
* PHP 256mb memory
* Memcached nodes in rhd.settings.php
* PHP version 7.2.10
* Set cache engine to 'memcached' in rhd.settings.php, rebuilt cache and then verified that Memcached nodes displayed as expected here: https://developer-preview-3009.ext.us-west.dc.preprod.paas.redhat.com/admin/reports/memcache and site functioned as expected
* The Grafana dashboard is display Memcache stats as expected

Thanks for all your hard work on this @robpblake !

**Comment by robpblake** [comment]
retest this please

**Pull request merged**